File: atcoder/ABC/abc201-300/abc270/c.py
from collections import deque
from collections import defaultdict
import sys
sys.setrecursionlimit(10000000)

input = sys.stdin.readline
n, x, y = map(int, input().split())

edges = defaultdict(lambda: [])
visited = [False for _ in range(n+1)]

for i in range(n-1):
    u_i, v_i = map(int, input().split())
    edges[u_i].append(v_i)
    edges[v_i].append(u_i)


q = deque()
q.append(x)

def dfs(current_edge, pre_edge):
    if current_edge == y:
        return True
    for next_edge in edges[current_edge]:
        if next_edge == pre_edge:
            continue
        q.append(next_edge)
        if dfs(next_edge, current_edge):
            return True
        q.pop()
    return False

dfs(x, -1)
print(*q)

File: atcoder/abc/abc201-300/abc270/c.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long int ll;
#define out cout
#define in cin

#define REP(i, n) for (ll i = 0, i##_len = (n); i < i##_len; ++i)

int main(){
    ll n, x, y;
    in >> n >> x >> y;
    map<ll, vector<ll>> edges = {};
    ll u, v;
    REP(i, n - 1){
        in >> u >> v;
        edges[u].push_back(v);
        edges[v].push_back(u);
    }

    deque<ll> dq = {};
    dq.push_back(x);
    bool visited[200001] = {false};

    while(1){
        ll count = 0;
        ll current_edge = dq[dq.size() - 1];
        visited[current_edge] = true;
        vector<ll> next_edges = edges[current_edge];

        REP(i, next_edges.size()){
            ll next_edge = next_edges[i];
            if(visited[next_edge]){
                continue;
            } else {
                dq.push_back(next_edge);
                count += 1;
                break;
            }
        }

        if(dq[dq.size() - 1] == y) break;
        if(count == 0){
            dq.pop_back();
        }
    }

    REP(i, dq.size()){
        out << dq[i];
        if(i == ll(dq.size() - 1)){
            out << endl;
        } else {
            out << " ";
        }
    }

    return 0;
}